synapse/storage/databases/main/roommember.py

```diff
@@ -687,19 +633,86 @@ def _get_users_server_still_shares_room_with_txn(
             _get_users_server_still_shares_room_with_txn,
         )
 
-    @cancellable
-    async def get_rooms_for_user(
-        self, user_id: str, on_invalidate: Optional[Callable[[], None]] = None
-    ) -> FrozenSet[str]:
+    @cached(max_entries=500000, iterable=True)
+    async def get_rooms_for_user(self, user_id: str) -> FrozenSet[str]:
         """Returns a set of room_ids the user is currently joined to.
 
         If a remote user only returns rooms this server is currently
         participating in.
         """
-        rooms = await self.get_rooms_for_user_with_stream_ordering(
-            user_id, on_invalidate=on_invalidate
+        rooms = self.get_rooms_for_user_with_stream_ordering.cache.get_immediate(
+            (user_id,),
+            None,
+            update_metrics=False,
+        )
+        if rooms:
+            return frozenset(r.room_id for r in rooms)
+
+        return await self.db_pool.runInteraction(
+            "get_rooms_for_user",
+            self._get_rooms_for_user_txn,
+            user_id,
+        )
+
+    def _get_rooms_for_user_txn(
+        self, txn: LoggingTransaction, user_id: str
+    ) -> FrozenSet[str]:
+        sql = """
+            SELECT room_id
+            FROM current_state_events AS c
+            WHERE
+                c.type = 'm.room.member'
+                AND c.state_key = ?
+                AND c.membership = ?
+        """
+
+        txn.execute(sql, (user_id, Membership.JOIN))
+        return frozenset(row[0] for row in txn)
+
+    @cachedList(
+        cached_method_name="get_rooms_for_user",
+        list_name="user_ids",
+    )
+    async def get_rooms_for_users(
+        self, user_ids: Collection[str]
+    ) -> Dict[str, FrozenSet[str]]:
+        """A batched version of `get_rooms_for_user`.
+
+        Returns:
+            Map from user_id to set of rooms that is currently in.
+        """
+        return await self.db_pool.runInteraction(
+            "get_rooms_for_users",
+            self._get_rooms_for_users_txn,
+            user_ids,
         )
-        return frozenset(r.room_id for r in rooms)
+
+    def _get_rooms_for_users_txn(
+        self, txn: LoggingTransaction, user_ids: Collection[str]
+    ) -> Dict[str, FrozenSet[str]]:
+
+        clause, args = make_in_list_sql_clause(
+            self.database_engine,
+            "c.state_key",
+            user_ids,
+        )
+
+        sql = f"""
+            SELECT c.state_key, room_id
+            FROM current_state_events AS c
+            WHERE
+                c.type = 'm.room.member'
+                AND c.membership = ?
+                AND {clause}
```

Comment:
Ditto, I think we can use `simple_select_many_batch` to simplify things.

Reply:
https://github.com/matrix-org/synapse/pull/13787/commits/21c8e8d69636c370df474226a81d8c85dc72d13d